File: package/main/src/String/slugify.ts
/**
 * Convert a string to a URL-friendly slug
 * @param str - The string to convert
 * @returns The slugified string
 * @example
 * slugify("Hello World!"); // "hello-world"
 * slugify("This is a Test"); // "this-is-a-test"
 * slugify("Japanese: こんにちは"); // "japanese"
 */
export const slugify = (string_: string): string => {
  return (
    string_
      .normalize("NFD")
      // Strip combining diacritical marks (e.g. accents)
      .replaceAll(/[\u0300-\u036F]/g, "")
      .toLowerCase()
      // Consolidate non-word chars, whitespace, and underscores into a single
      // hyphen in one pass instead of three separate regex scans
      .replaceAll(/[^\w-]+|_+/g, "-")
      // Collapse consecutive hyphens and strip leading/trailing hyphens
      .replaceAll(/-+/g, "-")
      .replaceAll(/^-|-$/g, "")
  );
};

File: package/umt_python/src/string/slugify.py
import re
import unicodedata


def slugify(string_: str) -> str:
    """
    Convert a string to a URL-friendly slug.

    Args:
        string_: The string to convert.

    Returns:
        The slugified string.

    Example:
        >>> slugify("Hello World!")
        'hello-world'
        >>> slugify("This is a Test")
        'this-is-a-test'
        >>> slugify("Café")
        'cafe'
    """
    result = unicodedata.normalize("NFD", string_)
    result = re.sub(r"[\u0300-\u036f]", "", result)
    result = result.lower()
    result = re.sub(r"[^\w\s-]", "-", result)
    result = re.sub(r"\s+", "-", result)
    result = re.sub(r"_+", "-", result)
    result = re.sub(r"-+", "-", result)
    return re.sub(r"^-+|-+$", "", result)
